### ISSUE TITLE
missing environmental variables nginx container

### ISSUE THREAD

**Issue by mpdevilleres** [pydanny/cookiecutter-django#626]
Hi,
I am trying to evaluate cookiecutter-django, so far its a great tool to start a project. everything is going smooth except when I tried to use docker with certbot, i'm having this error.

`nginx_1         | 2016/06/15 07:55:36 [emerg] 1#1: invalid port in upstream "___LETSENCRYPT_IP___:___LETSENCRYPT_PORT___" in /etc/nginx/nginx.conf:42
`

start.sh contains

```
sed -i "s/___LETSENCRYPT_IP___/$LETSENCRYPT_PORT_80_TCP_ADDR/g" /etc/nginx/nginx.conf
sed -i "s/___LETSENCRYPT_PORT___/$LETSENCRYPT_PORT_80_TCP_PORT/g" /etc/nginx/nginx.conf

```
I'm not a guru here, but I can see that the script is referencing with environmental variables, but when i tried to inspect the enviromental variables in nginx container, I didn't see anything related to letsencrypt.

```
[root@localhost]# docker-compose run nginx bash
Starting tbpcsuite_certbot_1
Starting tbpcsuite_postgres_1
Starting tbpcsuite_redis_1
Starting tbpcsuite_django_1
root@b7d12c245b86:/# set
BASH=/bin/bash
BASHOPTS=checkwinsize:cmdhist:complete_fullquote:expand_aliases:extquote:force_fignore:hostcomplete:interactive_comments:progcomp:promptvars:sourcepath
BASH_ALIASES=()
BASH_ARGC=()
BASH_ARGV=()
BASH_CMDS=()
BASH_LINENO=()
BASH_SOURCE=()
BASH_VERSINFO=([0]="4" [1]="3" [2]="30" [3]="1" [4]="release" [5]="x86_64-pc-linux-gnu")
BASH_VERSION='4.3.30(1)-release'
COLUMNS=197
DIRSTACK=()
EUID=0
GROUPS=()
HISTFILE=/root/.bash_history
HISTFILESIZE=500
HISTSIZE=500
HOME=/root
HOSTNAME=b7d12c245b86
HOSTTYPE=x86_64
IFS=$' \t\n'
LINES=73
MACHTYPE=x86_64-pc-linux-gnu
MAILCHECK=60
MY_DOMAIN_NAME=www.192.168.33.10.xip.io
NGINX_VERSION=1.11.1-1~jessie
OPTERR=1
OPTIND=1
OSTYPE=linux-gnu
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PIPESTATUS=([0]="1")
PPID=0
PS1='${debian_chroot:+($debian_chroot)}\u@\h:\w\$ '
PS2='> '
PS4='+ '
PWD=/
SHELL=/bin/bash
SHELLOPTS=braceexpand:emacs:hashall:histexpand:history:interactive-comments:monitor
SHLVL=1
TERM=xterm
UID=0
_=']'

```
I'm lost, any guidance I will appreciate.

Thanks,

**Comment by pydanny** [comment]
Hey @mjsisley, can you weigh in?

**Comment by mjsisley** [comment]
The environment variables are set by the certbot container and then used by the nginx container. This error likely means that the there is an issue with the certbot/ letsencrypt container. It is possible the command needs to change. 

Are there any logs for the certbot container prior to the "invalid port in upstream" error?

**Comment by pydanny** [comment]
The problem I had was that MY_DOMAIN_NAME environment variable was not set to the domain name I wanted.

**Comment by mjsisley** [comment]
So this is working for you now @pydanny?

**Comment by pydanny** [comment]
@mjsisley I had to get up a site quickly, so I did `Let's Encrypt` the manual way.

**Comment by pydanny** [comment]
Before we dig deeper into correcting this issue, I just wanted to compliment @mjsisley on the work he's done on `Let's Encrypt` and `Certbot` in Cookiecutter Django. He's solving a very hard problem, which is ensuring easy-to-use HTTPS for all users done through the abstraction that occurs when generating Cookiecutter templates.

With that, I also want to say that I expected a few bumps along the way, the same we encountered when we were polishing our Celery and Docker setups. These things happen, and it's perfectly okay. Especially in a framework like Cookiecutter Django which likes to push certain boundaries.

To summarize: Thank you @mjsisley for your hard work on the Certbot stuff. 💃

**Comment by mjsisley** [comment]
I like the idea that someone could essentially check an option on this project and then have HTTPS (thanks to the awesome thing that is Let's Encrypt). However, the manual way that Let's Encrypt provides is pretty easy as well.

Given that there is now an issue with the automated process that is in this project and that I now use [docker-letsencrypt-nginx-proxy-companion](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion) for Let's Encrypt, I think that one of the following options makes sense:

- Replace what is in the project with documentation for the manual way (or maybe Caddyfile).
- Try to find a way to use nginx-proxy and docker-letsencrypt-nginx-proxy-companion in a way that makes sense with this project structure (or could just have docs for how to use these).
- Something else that involves taking out what is in the project now (which worked for me for a few months, but apparently isn't working for others now)

**Comment by mirskiy** [comment]
I think this error may be related to docker's new networking features. Since the link environmental variables have been deprecated (https://docs.docker.com/compose/link-env-deprecated/), they are no longer replaced. I was able to get running by just using the actual container names in the conf, see #636. Seems it's already corrected for the django container - ___APPLICATION_IP___  has been replaced with django in the nginx confs.

**Comment by mpdevilleres** [comment]
This

**Comment by mpdevilleres** [comment]
This solves the problem, Thanks.

**Comment by aleprovencio** [comment]
It also solves for me, but as I mentioned on the PR, this new issue emerged:
``nginx: [emerg] host not found in upstream "certbot" in /etc/nginx/nginx.conf:42``

**Comment by aleprovencio** [comment]
It also solves for me, but as I mentioned on the PR, this new issue emerged:
``nginx: [emerg] host not found in upstream "certbot" in /etc/nginx/nginx.conf:42``

**Comment by mirskiy** [comment]
What version of docker and compose are you using? If you go into one of the containers, can you ping one of the other ones by name (ex. postgres)? I am on docker-compose version 1.7.1, build 0a9ab35, Docker version 1.11.2, build b9f10c9.

**Comment by aleprovencio** [comment]
Docker is the exact same version while docker-compose is almost (version 1.7.1, build 6c29830). Yes I can ping the container from the other ones, except the certbot which exits early since certificates are not yet due for renewal :(

**Comment by mirskiy** [comment]
Oh ok, seems like you might be getting the same error I was talking about in the PR. Have you already added `CMD /start.sh` back into the nginx Dockerfile? If so, I was able to get things working for now by removing the acme-challenge location blocks from nginx-secure.conf, it just means you will need to kill and restart the containers to perform certificate update, but hopefully there will be proper solution soon.

**Comment by aleprovencio** [comment]
Just to clear this up, you mean:

* Applying the changes from this PR
* put ``CMD /start.sh`` by the end of ``compose/nginx/Dockerfile``
* removing the two ``location /.well-known/acme-challenge`` blocks fom ``compose/nginx/nginx-secure.conf``

Is this correct?

**Comment by mirskiy** [comment]
Have you already performed the second step (changing the Dockerfile)? If not, I didn't think it would be failing - can you post your Dockerfile and the output that you are getting the issue in (just out of curiosity)?

Regardless, yes those steps are what I did to get it to work.

**Comment by aleprovencio** [comment]
@mirskiy, https seems to finally work via browser access!

Nginx however still threatens me throwing this on the logs:

	/start.sh: 54: kill: No such process
	[emerg] 19#19: bind() to 0.0.0.0:80 failed (98: Address already in use)

..and the Dockerfile:

	FROM nginx:latest
	ADD nginx.conf /etc/nginx/nginx.conf

	ADD start.sh /start.sh
	ADD nginx-secure.conf /etc/nginx/nginx-secure.conf
	ADD dhparams.pem /etc/ssl/private/dhparams.pem

	CMD /start.sh

**Comment by mirskiy** [comment]
The kill warning is probably due to grepping ps output returning the grep
process itself. I changed my start.sh kill line to `kill $(ps aux | grep
'nginx' | grep -v 'grep' | awk '{print $2}')
`. Or you can just ignore it, no difference.

The bind error could be because you already have something on port 80 on
your host machine. Are you running a web server on your host? If you are on
debian/ubuntu check the output of `sudo netstat -tulpn` without the
containers running and see if anything is listening on port 80.

**Comment by aleprovencio** [comment]
I think I'm ready to put on production my first cookiecutter/docker/letsencrypt setup!

@mirskiy, everything seems to be working now, thank you so much for you efforts!

And of course all you guys involved, you rock!

**Comment by lioman** [comment]
The certificates are generated but nginx container crashes, after certbot container is shut down.

**Comment by aleprovencio** [comment]
Even after doing everything I described here https://github.com/pydanny/cookiecutter-django/issues/626#issuecomment-227857208 ??

**Comment by jamespacileo** [comment]
We might have to keep this issue open, a patch needs to go in the repository :)

@aleprovencio Let me see if I got this right... we need the following:

- `CMD /start.sh` in the Dockerfile

- Explicitly add all environment variables pertaining the django and certbot containers, as compose v2 doesn't support auto-generated envs anymore

**Comment by aleprovencio** [comment]
Here you go:

* CMD /start.sh in the Dockerfile
* remove the `location /.well-known/acme-challenge` blocks from nginx-secure.conf
* make the changes for nginx.conf and start.sh provided [here](https://github.com/pydanny/cookiecutter-django/pull/636/files)

**Comment by mirskiy** [comment]
@jamespacileo You can use the names of the containers instead of the env. vars (see #636 ) - docker's new networking handles the name resolution.

The problem is nginx isn't happy if one of the upstreams isn't up. After we run with the initial nginx.conf and run certbot, we restart with the nginx-secure.conf but certbot is no longer up. Since nginx-secure.conf references certbot (and it needs to in order to have renewals work), nginx can't switch to the new conf.

I've updated the original pull request (#636) with a better workaround that should support renewals. By setting the proxy_pass server as a variable and using the resolver directive, I'm telling nginx to perform the container name -> resolution dynamically. I'm getting the dns server of the container from /etc/resolv.conf (this will point at the docker network dns server) and setting the resolver to it.

Based on:
https://trac.nginx.org/nginx/ticket/658
https://tenzer.dk/nginx-with-dynamic-upstreams/
http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_pass
